include/quic/parser.hpp

```diff
@@ -0,0 +1,421 @@
+#include <oxenc/bt.h>
+
+#include "stream.hpp"
+#include "utils.hpp"
+
+namespace oxen::quic
+{
+    inline auto bp_cat = oxen::log::Cat("bparser");
+
+    using time_point = std::chrono::steady_clock::time_point;
+
+    // timeout is used for sent requests awaiting responses
+    const std::chrono::seconds TIMEOUT{10};
+
+    // request sizes
+    inline constexpr long long MAX_REQ_LEN = 10_M;
+
+    // Application error
+    inline constexpr uint64_t BPARSER_EXCEPTION = (1ULL << 60) + 69;
+
+    class bparser;
+
+    struct message
+    {
+        friend class bparser;
+
+      private:
+        int64_t req_id;
+        std::string data;
+        std::string_view req_type;
+        std::string_view endpoint;
+        std::string_view req_body;
+        std::weak_ptr<bparser> return_sender;
+        bool timed_out{false};
+
+      public:
+        message(bparser& bp, std::string req, bool is_error = false);
+
+        void respond(int64_t rid, std::string body, bool error = false);
+
+        //  To be used to determine if the message was a result of an error as such:
+        //
+        //  void f(const message& m)
+        //  {
+        //      if (not m.timed_out)
+        //      { // success logic }
+        //      ... // is identical to:
+        //      if (m)
+        //      { // success logic }
+        //  }
+        operator bool() const { return not timed_out; }
+
+        int64_t rid() { return req_id; }
+        std::string_view view() { return {data}; }
+    };
+
+    struct sent_request
+    {
+        // parsed request data
+        int64_t req_id;
+        std::string data;
+        bparser& return_sender;
+
+        // total length of the request; is at the beginning of the request
+        size_t total_len;
+
+        std::chrono::steady_clock::time_point req_time;
+        std::chrono::steady_clock::time_point timeout;
+
+        bool is_empty() const { return data.empty() && total_len == 0; }
+
+        explicit sent_request(bparser& bp, std::string d, int64_t rid) : req_id{rid}, return_sender{bp}
+        {
+            total_len = d.length();
+            data.reserve(data.length() + total_len);
```

Comment:
I think just drop this reserve; it isn't quite right, and the "right" value here is annoyingly difficult to calculate (because it involves stringification of an integer).  But small-string-optimization will most likely save us anyway since the initial `"1234:"` will get written before it allocates anything.

Reply:
(I think the "correct" calculation is `1 + ceil(log10(total_len)) + d.size()`, but to hell with calculating that).